### PR TITLE
Add timestamp and timezone management documentation

### PR DESCRIPTION
**Pull request by pedroerp**
Adding a new documentation page to describe details of management of timestamps, timezone and conversions in Velox.

### CODE REVIEW

**Comment by pedroerp** [comment]
Cc: @PHILO-HE

File: velox/docs/develop/timestamp.rst [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+<https://github.com/facebookincubator/velox/blob/main/velox/type/Timestamp.h>`_
+class. The Timestamp class stores two 64 bit integers, one containing the
+number of seconds from ``1970-01-01 00:00:00``, and one containing the
+nanoseconds offset in that particular second, in order to provide nanosecond
+precision. A few more observations:
+
+* While “seconds” can be negative to represent time before ``1970-01-01
+  00:00:00``, “nanoseconds” are always positive.
+
+* Engines like Presto and Spark only support milliseconds precision, so
```
Comment (mbasmanova) [inline velox/docs/develop/timestamp.rst:61]: I believe Spark supports at least microseconds precision.

File: velox/docs/develop/timestamp.rst [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+  naturally “nanoseconds” are zeroed.
```
Comment (mbasmanova) [inline velox/docs/develop/timestamp.rst:62]: Nanoseconds are used even with millisecond precision.

File: velox/docs/develop/timestamp.rst [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+Timezone IDs in Velox are based on the id map used by Presto and are 
+`available here <https://github.com/prestodb/presto/blob/master/presto-common/src/main/resources/com/facebook/presto/common/type/zone-index.properties>`_. 
+They are automatically generated using `this script <https://github.com/facebookincubator/velox/blob/main/velox/type/tz/gen_timezone_database.py>`_. 
+While timezone IDs are an implementation detail and ideally should not leak
+outside of Velox execution, they are exposed if data containing
+TimestampWithTimezones are serialized, for example.
+
+**TimestampWithTimezone:** To represent an absolute point in time, Velox provides
+the `TimestampWithTimezone class <https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/types/TimestampWithTimeZoneType.h>`_.
```
Comment (mbasmanova) [inline velox/docs/develop/timestamp.rst:84]: I don't think there is a TimestampWithTimezone class.

File: velox/docs/develop/timestamp.rst [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+This class implements the TIMESTAMP WITH TIMEZONE sql semantic discussed above,
```
Comment (mbasmanova) [inline velox/docs/develop/timestamp.rst:85]: sql -> SQL

File: velox/docs/develop/timestamp.rst [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+and was based on Presto’s implementation - therefore only supporting
```
Comment (mbasmanova) [inline velox/docs/develop/timestamp.rst:86]: was based -> is based

File: velox/docs/develop/timestamp.rst [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+millisecond-precision.
+
+TimestampWithTimezone physically packs two integers in a single 64 word, using
+12 bits for timezone ID, and 52 for a millisecond-precision timestamp.
```
Comment (mbasmanova) [inline velox/docs/develop/timestamp.rst:90]: 52 -> 52 bits

File: velox/docs/develop/timestamp.rst [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+This class handles timezone conversions by leveraging APIs provided by the
+operating system, based on the tzdata database installed locally. If systems
+happen to have inconsistent or older versions of the tzdata database, Velox’s
+conversions may produce inconsistent results. 
+
+You can check the tzdata installed in your system by:
+
+.. code-block:: bash
+
+  $ rpm -qa | grep tzdata
```
Comment (mbasmanova) [inline velox/docs/develop/timestamp.rst:132]: I tried this command on my Mac, but it didn't print anything.

**Review by mbasmanova — APPROVED** [review]
Thank you for putting this together. Super useful. Some nits.

File: velox/docs/develop/timestamp.rst [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+semantics, using `Presto as a reference implementation <https://prestodb.io/docs/current/functions/datetime.html>`_,
+using ``set session legacy_timestamp = false;`` (see the section below for
+details).
+
+Timestamp literals are created based on whether time zone information is found
+on the string on not:
+
+.. code-block:: sql
+
+  presto:de> select typeof(TIMESTAMP '1970-01-01 00:00:00');
```
Comment (zacw7) [inline velox/docs/develop/timestamp.rst:155]: Let's omit the `presto:de` part? Should we keep the format of the examples consistent as https://facebookincubator.github.io/velox/functions/presto/conversion.html#cast-to-timestamp

**Review by zacw7 — APPROVED** [review]
Thanks for composing this doc. This is super helpful!

File: velox/docs/develop/timestamp.rst (hunk 2 above)
Comment (pedroerp) [reply velox/docs/develop/timestamp.rst:62]: Rephrased it

File: velox/docs/develop/timestamp.rst (hunk 3 above)
Comment (pedroerp) [reply velox/docs/develop/timestamp.rst:84]: well, it's an alias:

https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/types/TimestampWithTimeZoneType.h

I'll rephrase.

File: velox/docs/develop/timestamp.rst (hunk 7 above)
Comment (pedroerp) [reply velox/docs/develop/timestamp.rst:132]: this is for linux; added to the text to clarify

File: velox/docs/develop/timestamp.rst (hunk 8 above)
Comment (pedroerp) [reply velox/docs/develop/timestamp.rst:155]: Nice! That's better

**Pull request closed without merging**